### DISCUSSION TITLE
Setup JWT Manual Clerk.net with .Net 8 Web API

### DISCUSSION THREAD

**Discussion by ttaylor29** [Hawxy/Clerk.Net#6] [General]
Hello,

I am new to Clerk.com. I'm trying to start a next.js app with Clerk.

I want to do all of my backend with .NET Core WEB API.

I figured out tonight how to do a manual JWT.

I want to just basically use the front end for the user to sign in and then use that token/JWT to verify they are authenticated. 

I tried to set up your code and for some reason, I have a 401 not authorized when I make the call from my next.js app.

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/851eb6f7-765c-41d2-9b39-4969ad5d966a)

I would rather use your SDK for .NET compared to checking manual JWTs and then using the userId from the principle to go and query their API and setting the claims.

Do you know what I'm doing wrong by chance?

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/c7de08b9-c7bf-4685-8fc2-71e4ac24b08f)

Here is my .tsx file:

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/853f7d37-f878-4330-9628-606627d259d4)

**Comment by Hawxy** [comment]
The `ClerkAuthority` URL you've provided likely isn't valid. Make sure your appsettings.json is correct.

**Reply by ttaylor29** [reply]
@Hawxy  Let me try this.

**Reply by ttaylor29** [reply]
This is what I have:

  "ClerkAuthority": "https://X.clerk.accounts.dev",
  "ClerkAuthorizedParty": "http://localhost:3000",

Is this right?

It seems like it is pulling these values in, but nothing happens when I navigate to the web api.

**Reply by ttaylor29** [reply]
@Hawxy 

I was not setting the  **IssuerSigningKey = new RsaSecurityKey(rsa),** part.

Here is my code:

Program.cs section to add JWT items:

```
                    builder.Services.AddClerkApiClient(config =>
                    {
                        config.SecretKey = builder.Configuration["ClerkSecretKey"];
                    });


                    builder.Services.AddJwtBearerAuthentication(builder.Configuration, builder.Services.BuildServiceProvider().GetRequiredService<IHttpClientFactory>());

```

Here is a static file I made. This sets the claims for me as well. 

```
using Microsoft.AspNetCore.Authentication.JwtBearer;
using Microsoft.IdentityModel.Tokens;
using System.Net.Http.Headers;
using System.Security.Claims;
using System.Security.Cryptography;
using System.Text.Json;

namespace WebAPINext.Configuration
{
    public static class JwtBearerAuthenticationConfiguration
    {
        public static void AddJwtBearerAuthentication(
            this IServiceCollection services, IConfiguration configuration, IHttpClientFactory httpClientFactory)
        {
            services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
                .AddJwtBearer(options =>
                {
                    // Assume publicKey is your Clerk instance's public RSA key in PEM format
                    var pemFormat = configuration["ClerkPublicKeyPemFormat"];

                    if (pemFormat == null)
                    {
                        throw new Exception(nameof(pemFormat));
                    }

                    RSA rsa = RSA.Create();
                    rsa.ImportFromPem(pemFormat.ToCharArray());

                    options.Authority = configuration["ClerkAuthority"];
                    options.TokenValidationParameters = new TokenValidationParameters
                    {
                        ValidateAudience = false,
                        ValidateIssuerSigningKey = true,
                        ValidateLifetime = true,
                        IssuerSigningKey = new RsaSecurityKey(rsa),
                        ValidateIssuer = false,
                        ClockSkew = TimeSpan.Zero
                    };
                    options.Events = new JwtBearerEvents
                    {
                        OnTokenValidated = async context =>
                        {
                            var azp = context.Principal?.FindFirstValue("azp");
                            if (string.IsNullOrEmpty(azp) || !azp.Equals(configuration["ClerkAuthorizedParty"]))
                            {
                                context.Fail("AZP Claim is invalid or missing");
                                return;
                            }

                            await SetClaims(context, configuration, httpClientFactory).ConfigureAwait(false);
                        },
                    };
                });
        }

        private static async Task SetClaims(TokenValidatedContext context, IConfiguration configuration, IHttpClientFactory httpClientFactory)
        {
            try
            {
                // Assuming 'configuration' and 'context' are available in the scope
                var clerkApiKey = configuration["ClerkSecretKey"];
                if (string.IsNullOrEmpty(clerkApiKey))
                {
                    throw new InvalidOperationException("Clerk API key is not configured properly.");
                }

                var userIdClaim = context.Principal.Claims.FirstOrDefault(c => c.Type == "user_id" || c.Type == ClaimTypes.NameIdentifier);
                if (userIdClaim == null || string.IsNullOrEmpty(userIdClaim.Value))
                {
                    throw new InvalidOperationException("User ID claim is not available.");
                }

                string userId = userIdClaim.Value;
                string clerkApiUrl = $"https://api.clerk.com/v1/users/{userId}";

                using var client = httpClientFactory.CreateClient();
                client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", clerkApiKey);

                var response = await client.GetAsync(clerkApiUrl);
                if (!response.IsSuccessStatusCode)
                {
                    Console.WriteLine("Failed to retrieve user details from Clerk.");
                    return;
                }

                var responseBody = await response.Content.ReadAsStringAsync();
                using var jsonDoc = JsonDocument.Parse(responseBody);

                var user = jsonDoc.RootElement;

                var id = user.GetProperty("id").GetString();
                var emailAddress = user.GetProperty("email_addresses")[0].GetProperty("email_address").GetString();
                var firstName = user.GetProperty("first_name").GetString();
                var lastName = user.GetProperty("last_name").GetString();
                var userName = user.GetProperty("username").GetString();

                var claims = new List<Claim>
                {
                    new Claim(ClaimTypes.NameIdentifier, id),
                    new Claim(ClaimTypes.Name, userName),
                    new Claim("firstName", firstName),
                    new Claim("lastName", lastName),
                    new Claim(ClaimTypes.Email, emailAddress),
                };

                if (context.Principal.Identity is ClaimsIdentity claimsIdentity)
                {
                    claimsIdentity.AddClaims(claims);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in VerifyTokenWithClerkAsync: {ex.Message}");
            }
        }
    }
}

```

**Reply by ttaylor29** [reply]
I'm sure this can be cleaned up  more in the future.

**Reply by Hawxy** [reply]
None of that code should be needed, ASP.NET Core queries Clerk's JWKs and grabs the public key itself via `https://*.clerk.accounts.dev/.well-known/jwks.json`. I just retested this now and my example works fine.

**Comment by ttaylor29** [comment]
Well, interesting.

I'm not sure what else it is?

I verified the Authority item.

Do you have a test Web Api .Net 8?

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by Hawxy** [reply]
You might just be missing a forward slash at the end of your authority URL. You've disabled issuer validation to get things working, which means you aren't validating the authority at all.

Check a JWT on https://jwt.io/ to make sure it's lined up.

Your set claims code also isn't needed, you can modify the JWT claims via "Sessions -> Customize session token" on the Clerk dashboard.

**Reply by Hawxy** [reply]
I'd suggest debugging your code to make sure the Authority is being loaded from your configuration and set as expected.

**Reply by ttaylor29** [reply]
@Hawxy It seems correct?

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/e81de306-7939-4ee0-9f62-fdaff8b961e9)

Here is the latest setup

```

            services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
                .AddJwtBearer(options =>
                {
                    var clerkAuthority = configuration["ClerkAuthority"];

                    options.Authority = clerkAuthority;
                    options.TokenValidationParameters = new TokenValidationParameters()
                    {
                        // Disable audience validation as we aren't using it,
                        ValidateAudience = false,
                        NameClaimType = ClaimTypes.NameIdentifier
                    };
                    options.Events = new JwtBearerEvents
                    {
                        OnTokenValidated = async context =>
                        {
                            var azp = context.Principal?.FindFirstValue("azp");
                            if (string.IsNullOrEmpty(azp) || !azp.Equals(configuration["ClerkAuthorizedParty"]))
                            {
                                context.Fail("AZP Claim is invalid or missing");
                                return;
                            }
                        },
                    };
                });

```

**Reply by ttaylor29** [reply]
(I'll look into setting the claims via the dashboard later after I figure out why I get a 401 when I use the setup from your setup)

**Reply by ttaylor29** [reply]
@Hawxy 

What does this mean?

Www-Authenticate:
Bearer error="invalid_token", error_description="The signature key was not found"

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/299f7dcd-d377-443a-955a-435730dafa7a)

It's never getting to the `OnTokenValidated ` part either.

**Reply by Hawxy** [reply]
Just to make sure, you have the latest version of `Microsoft.AspNetCore.Authentication.JwtBearer` installed and this is .NET 8?

**Reply by ttaylor29** [reply]
Yes, this is .NET 8.

Here is the latest version 8.0.3

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/bb6f52f0-ab15-4b13-9b99-b36eb18743a2)

**Reply by Hawxy** [reply]
Alright. In your appsettings, set your logging level to `Trace`. When you hit your backend you'll get the full exception from the authentication middleware.

**Reply by Hawxy** [reply]
Could you also check that the below two urls work in your browser:
```
https://{clerksubdomain}.clerk.accounts.dev/.well-known/openid-configuration
https://{clerksubdomain}.clerk.accounts.dev/.well-known/jwks.json
```

**Reply by ttaylor29** [reply]
Here is:

https://{clerksubdomain}.clerk.accounts.dev/.well-known/openid-configuration

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/2a51791c-8d25-4364-bbfa-fa5e1aa60557)

**Reply by ttaylor29** [reply]
Here is:  https://{clerksubdomain}.clerk.accounts.dev/.well-known/jwks.json

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/93f972cf-6d7f-407f-9fae-8136b0970f29)

**Reply by ttaylor29** [reply]
> Alright. In your appsettings, set your logging level to Trace. When you hit your backend you'll get the full exception from the authentication middleware.

Where does it log those errors? I have it set to Trace? I do not see anything on the command line console?

```
{
  "Logging": {
    "LogLevel": {
      "Default": "Trace",
      "Microsoft.AspNetCore": "Trace"
    }
  },
  "AllowedHosts": "*"
}

```

**Reply by Hawxy** [reply]
It would be in the console yes with the rest of your logs.

I don't really get what's going on with your project. I've pushed up a sample that I've validated works fine against my own clerk instance:
https://github.com/Hawxy/Clerk.Net/tree/main/src/Clerk.Net.JwtSample

**Reply by ttaylor29** [reply]
Maybe it is the Authority.

Does this look like the correct Authority?

https://{clerkSubDomain}.clerk.accounts.dev/

or 

https://{clerkSubDomain}.accounts.dev/

Can you send me a screenshot where you get this from the dashboard?

I can also try your example you made later to see if I get the same thing.

**Reply by ttaylor29** [reply]
Also,

Yea, I don't see anything in the logs of the web api that are showing up:

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/46d462ad-014e-4a35-adfe-daec7eb0eb2a)

**Reply by Hawxy** [reply]
`Developers > Domains`. You can validate this by checking the `iss` value of a JWT via https://jwt.io/

**Reply by Hawxy** [reply]
Did you set it in `appsettings.Development.json`? That will take precedent over the other appsettings when running locally.

**Reply by ttaylor29** [reply]
Well I'm saving them in my secret.json?

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/9077d6bf-6934-4b13-9488-f83ca646b025)

**Reply by Hawxy** [reply]
I'm referring to the logging configuration.

**Reply by ttaylor29** [reply]
Oh okay, wait a second. I don't have that file.

Also here is the F12 errors.

Maybe my react app isn't even getting to the web api even though it is listening.

What should happen if I go directly in the browser?

It isn't doing anything at that point either:

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/28130310-d3e4-46f0-b904-dc71693ff6a7)

![image](https://github.com/Hawxy/Clerk.Net/assets/23564255/b5ede749-cd31-4e27-96a4-423db8c35faa)

**Reply by ttaylor29** [reply]
I would think that even via Chrome it would work and not even via my next.js app.

**Reply by Hawxy** [reply]
All of that is expected behavior.

**Reply by Hawxy** [reply]
There's no bearer attached so naturally the backend is just gonna 401 if you try to access the URL.

**Reply by Hawxy** [reply]
Check the appsettings.json being output in the bin/debug folder, you'll probably find the logging settings weren't updated.

**Reply by ttaylor29** [reply]
Oh, so maybe the secret.json is not being pushed after it builds?

**Reply by Hawxy** [reply]
No, that's not how the `secret.json` works. It gets read directly, doesn't get output.

**Reply by ttaylor29** [reply]
How about this:

dbug: Microsoft.AspNetCore.Cors.Infrastructure.CorsService[2]
      The request has an origin header: 'http://localhost:3000'.
info: Microsoft.AspNetCore.Cors.Infrastructure.CorsService[4]
      CORS policy execution successful.
info: Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler[1]
      Failed to validate the token.
      Microsoft.IdentityModel.Tokens.SecurityTokenSignatureKeyNotFoundException: IDX10500: Signature validation failed. No security keys were provided to validate the signature.
         at Microsoft.IdentityModel.JsonWebTokens.JsonWebTokenHandler.ValidateSignature(JsonWebToken jwtToken, TokenValidationParameters validationParameters, BaseConfiguration configuration)
         at Microsoft.IdentityModel.JsonWebTokens.JsonWebTokenHandler.ValidateSignatureAndIssuerSecurityKey(JsonWebToken jsonWebToken, TokenValidationParameters validationParameters, BaseConfiguration configuration)
         at Microsoft.IdentityModel.JsonWebTokens.JsonWebTokenHandler.ValidateJWSAsync(JsonWebToken jsonWebToken, TokenValidationParameters validationParameters, BaseConfiguration configuration)
info: Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler[7]

**Reply by ttaylor29** [reply]
Now I'm seeing logs though in my console after I moved the `secret.json` items to my `appsettings.Development.json` file.

**Reply by Hawxy** [reply]
Yeah it's as I thought, there's no JWKs being loaded and thus no public key to check against. I've never seen this type of behavior before.

I'd suggest cloning my sample, adding the basic config and seeing if your frontend can hit it, just to rule out something being odd in your own project.

**Reply by ttaylor29** [reply]
I found it!!!!!!!!!!!!!!!!!!!!!!!!

Look at this URL: https://stackoverflow.com/a/78200588

I did what they said (removed those references that somehow I added via another process, I guess (maybe OAuth setup)).

NOW YOUR CODE WORKS PERFECT!!!

I need to add my claims via the website so I don't have to return to get them via an API call.

**Reply by Hawxy** [reply]
Man that's very unfortunate. I'm surprised having both installed caused such a conflict. Glad you worked it out though as I was effectively out of ideas.

**Reply by ttaylor29** [reply]
@Hawxy  Thanks again for your help! Not saying I won't have more questions as I become more familiar with NextJS / Clerk.com / React with your awesome product!

**Comment by appwebcaddy** [comment]
Few things that was not clear in the docs are:
1. In order to get the token, you need to create a JWT Template, this is in Clerk Dashboard. simple default empty template will do.
2. check the token https://jwt.io/ make sure is valid
3. docs says: "If the `azp` claim does not exist, then you can skip this step." so `if (string.IsNullOrEmpty(azp) || )... error` should not return an error. 
https://clerk.com/docs/backend-requests/handling/manual-jwt#verify-the-token-signature
4. if not sure about what is your clerk Authority, go to Clerk Dashboard > Developers > API Keys > Show API URLs > Frontend API URL

**Reply by Hawxy** [reply]
> In order to get the token, you need to create a JWT Template

This isn't true at all. Clerks session token (which is a JWT) works fine. It can be expanded via the customization options if needed.

> check the token https://jwt.io/ make sure is valid

Sure, that's useful to do, but I'm assuming a user has at least a basic understanding of JWTs in my example.

> docs says: "If the azp claim does not exist, then you can skip this step."

If you're using the Clerk JS SDK within a SPA to generate tokens then the azp claim should always be present, which is what I would consider the common usage of having an ASP.NET Core backend. If azp is missing, your usage is unusual.

**Reply by appwebcaddy** [reply]
@Hawxy this how I generate the token using clerk with NextJs... I am not sure why the api required the template to generate the token, and why the azp is not present? notice I am using a new free clerk account in dev mode. 
Can you provide us with an example/doc link, how you generate your JWT token?
![image](https://github.com/Hawxy/Clerk.Net/assets/134953582/559ea7e5-290b-414b-9c14-2760cd46e78c)

**Reply by Hawxy** [reply]
I use the front-end session token (I personally don't use JS SSR frameworks, I typically use a SPA/static site).  If you're sending requests from Next SSR, then you'll want to read the `__session` cookie to grab the JWT (afaik)
